### ISSUE TITLE
Add a variable in twisted.python.runtime to check for vista

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2881]
|<img alt="bigdog's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| bigdog reported|
|-|-|
|Trac ID|trac#2881|
|Type|defect|
|Created|2007-11-11 16:45:04Z|
|Branch|https://github.com/twisted/twisted/tree/vista-2881|

os.name returns 'nt' and sys.platform returns 'win32', so it's impossible to differentiate vista from XP with that.  Maybe we could extract the information from ```platform.uname()[3]``` (whose value is 
'6.0.6000' in vista).

See the related bug in Python here: http://bugs.python.org/issue1082.

Attachments:

* [2881.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/8f6/8f66fb2c5891725f0f13699624db23a610c94764/ee5c7cd3da4535249d55c6dc56e3dcadc560bdc1.diff) (1297 bytes) - added by therve on 2008-05-10 11:00:34Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__2881 2881
type__defect defect
reporter__bigdog bigdog
priority__high high
milestone__ 
branch__branches_vista_2881 branches/vista-2881
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1194799504000000 1194799504000000
changetime__1211894779000000 1211894779000000
version__None None
owner__ 
cc__glyph
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph commented|
|-|-|

I am sorely tempted to close this as invalid.  Why does anyone ever need to know the difference between XP and Vista using Twisted?  This seems to me like modifying twisted.python.runtime to have different values to check for Ubuntu 6.06 vs. 7.10.  The only consequence I can see is adding a new way for code to ask "is this a windows machine" that will succeed on XP but fail on Vista (or vice versa).  Shouldn't we leave platform detection of this granularity up to the code which will actually care?  i.e. if you are asking if it's Vista because Vista has a new or changed win32 function, shouldn't you be asking the win32 wrapper?

If Twisted needs to make this determination for its own purposes, then this makes sense to me.  Is Twisted going to be making different API calls on Vista and XP?  If not, I'd like to consider them the same platform.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve commented|
|-|-|

Again, sorry for not being clear, I'm the one who asked for that. This may be something we want because vista UDP implementation is significantly different from XP one (for example, test_joinFailure seems to pass under vista).

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve commented|
|-|-|

(In [23533]) Branching to 'vista-2881'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve removed owner|
|-|-|

OK this is ready to review. This branch is green under my vista instance. I also realized that buildbot wrongly reports successful build when we have unexpected success (like on the windows 2008 slave).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to @therve|
|-|-|

Looks good, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve set status to `closed`|
|-|-|

(In [23691]) Merge vista-2881

Author: therve
Reviewer: exarkun
Fixes [#2881](https://github.com/twisted/twisted/issues/2881)

Add twisted.python.runtime.Platform.isVista: this method is used to
differentiate Vista/Server 2008 from previous Windows versions. This is first
used to change a Multicast test from todo to pass.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve set status to `reopened`|
|-|-|

(In [23693]) Revert r23691: regressions under OS X. Some failures like that:

```
[ERROR]: twisted.python.test.test_runtime.PlatformTests.test_isVistaConsistency

Traceback (most recent call last):
  File "/Users/buildbot/Run/buildbot/full2.5-exarkun-osx/Twisted/twisted/python/test/test_runtime.py", line 26, in test_isVistaConsistency
    if platform.isVista():
  File "/Users/buildbot/Run/buildbot/full2.5-exarkun-osx/Twisted/twisted/python/runtime.py", line 85, in isVista
    uname = platform.uname()
  File "/System/Library/Frameworks/Python.framework/Versions/2.5/lib/python2.5/platform.py", line 1014, in uname
    processor = _syscmd_uname('-p','')
  File "/System/Library/Frameworks/Python.framework/Versions/2.5/lib/python2.5/platform.py", line 800, in _syscmd_uname
    output = string.strip(f.read())
exceptions.IOError: [Errno 4] Interrupted system call
```

Reopens [#2881](https://github.com/twisted/twisted/issues/2881)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve commented|
|-|-|

(In [23699]) Replace platform.uname by sys.getwindowsversion

Refs [#2881](https://github.com/twisted/twisted/issues/2881)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_11"></a>|@therve removed owner<br>@therve set status to `new`|
|-|-|

That looks better.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set owner to @therve|
|-|-|

I guess we can probably just use `sys.getwindowsversion` instead of that and `win32api.GetVersionEx`.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_13"></a>|@therve removed owner|
|-|-|

That's a good guess.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun set owner to @therve|
|-|-|

Looks good... again. ;)  Please merge.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_15"></a>|@therve set status to `closed`|
|-|-|

(In [23715]) Merge vista-2881

Author: therve
Reviewer: exarkun
Fixes [#2881](https://github.com/twisted/twisted/issues/2881)

Add twisted.python.runtime.Platform.isVista: this method is used to
differentiate Vista/Server 2008 from previous Windows versions. This is first
used to change a Multicast test from todo to pass.

(Remerge using sys.getwindowsversion instead of platform.uname)

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
